### PR TITLE
feat(DualListSelector): add support for trees and checkboxes

### LINKED ISSUE #5102: Dual list selector - add menu enhancements
Add enhancements to support collapsible tree in lists per: https://github.com/patternfly/patternfly/issues/3615

Should mimic behaviors of the existing Tree component.

### PR DESCRIPTION
**Pull request by kmcfaul**
**What**: Closes #5102 

To swap over to using trees, add `isTree` to the `DualListSelector` and update the data structure accordingly. The data is expected to have a `text` attribute to display the content of the item, and this property is used to control selection/checking/etc. The structure is similar to TreeView, but uses `text` in place of `name`. It might be better to swap to using an `id` prop instead of `text`, I haven't decided.

The prop `hasChecks` added to `DualListSelector` will add checkboxes to every option. Alternately, adding a `hasCheck` property to an option will give the individual option a checkbox. Badges are added individually at the moment with a `hasBadge` property added to an option, but if needed a similar apply-to-all property can be added.

Keyboard interaction will require some thought, and maybe requires a follow up issue, because TreeView used tabbing to navigate and DualListSelector skips over tabbing through the options.

### CODE REVIEW

**Comment by mcarrano** [comment]
@mmenestr can you take a look at this? Not sure exactly how this should be working relative to your design. Thanks.

**Comment by kmcfaul** [comment]
I tried to follow TreeView mostly, but had to make some assumptions about how things like the checkboxes would behave in a DualListSelector context (as selectors instead of being separate in TreeView), so any feedback would be great on interaction.

**Comment by mmenestr** [comment]
It looks pretty good! The only thing I notice is that on nodes with no expansion, clicking on the item puts it into a selection state, but doesn't check the checkbox. It should either check the checkbox AND put the whole thing in a selected state, or just check the checkbox, and keep the text and background the same. Basically, there's a weird intermediate state that shouldn't exist! See image:
image

The other thing i'm iffy about is the fact that clicking on a row **with** an expansion both selects it AND opens it. I don't think that should be the behavior and I wonder if we could do one of two things:

**Option 1:** Separate the row into two "click areas", and make it so that if they click anywhere left of the checkbox - so on the arrow or left of it, the row expands, but if they click on the checkbox or anything right of the checkbox, then it selects the item without expanding? 

**Option 2:** Decide that users can't select something unless they click on the checkbox itself, and any other click on the row would either expand the row (for rows with carets) or do nothing (for rows without). This might be the simplest and more intuitive option.

Curious to hear your thoughts on what you'd expect^ @mcarrano @kmcfaul

**Comment by kmcfaul** [comment]
I think that if the user wants the checkboxes, it would make sense for them to replace the row selection. At least I'm not sure if there's a use case for checkboxes outside of selection.

I can implement either way though. I already separated expansion from selection for the toggle and can do the same for the whole row (so expansion is controlled only by the toggle button, not the row click).

**Comment by kmcfaul** [comment]
Realized I didn't update the search functionality with trees, so working on including that as well as a basic integration demo/test.

**Comment by nicolethoen** [comment]
This is cool! I'd say that if they are using checkboxes, then possibly the checkboxes should just toggle alongside the selected state? so the checkbox just emphasizes the selected state?
and a partially checked checkbox would not imply the selected state?

**Comment by mmenestr** [comment]
@nicolethoen I think that makes sense! 
And as for the expansion vs selection thing, I just got a second designer opinion and we're both leaning towards keeping it simple and going with option 2 which would be:
-  if the user clicks on the checkbox, the row gets selected
- if the user clicks anywhere else (arrow or general row outside of checkbox) the folder expands. 
- But there should never be a place where a click causes both a selection AND a folder expansion

**Comment by mcarrano** [comment]
Agree with @mmenestr above. I think that this also means that clicking outside of the checkbox on a row without children has no effect (the user must use the checkbox to select/deselect items).

**Comment by kmcfaul** [comment]
Updated! Whenever an option has a checkbox, it will prevent the row selection and only allow the checkbox toggle. So if a user is using the tree, but not checkboxes, row selection will behave as before.

The checkbox handler in DualListSelector handles the checkbox as well as selection, and also now triggers the optional selection callback.

**Comment by mcarrano** [comment]
The interaction feels right now @kmcfaul . Do you know if this is consistent with how we are handling the Tree view? It should behave the same, I think. I'm also wondering if it's possible to clean up the alignment issue I'm seeing below:

Screen Shot 2020-12-03 at 4 04 24 PM

Does that require a core fix?

Also, I'm not clear what the purpose of the badge is. Can you remind me of what that is intended to indicate @mmenestr ? It doesn't seem to change with anything.

**Comment by kmcfaul** [comment]
The alignment of folders to options would probably be a core issue. @mcoker 

The badge is showing the number of direct children. This is the same as Tree View and I think what the core example showed, but I'm unsure.

It's not the same interaction, TreeView's selection and checking behaviors were separate. Selection only highlighted the option and its direct parent, while checking could be done to any option and didn't trigger selection at all. I think the context of Dual List Selector changes the behavior opposed to Tree View (which may be used for a menu or the like). But I could be wrong, should something change?

**Comment by tlabaj** [comment]
@mcarrano @mmenestr wouldn't we want this component's interactions to be consistent with the tree view component.  See PR #https://github.com/patternfly/patternfly-react/pull/5128 for the updates to the checkbox interactions in the tree view component. 
The only way to expand the node is the click the toggle.  clicking the label or checkbox selects the item.

**Comment by mcarrano** [comment]
@tlabaj I had the same thought, initially. But @kmcfaul makes a good point that the usage context is a little bit different. What do you think @mmenestr ? I can live with the inconsistency if we think that both components make sense in their separate contexts. Users won't think hard about this if it just works.

**Comment by mmenestr** [comment]
Looks a lot better! Just have 2 things:
1. I'm still able to click on an item row, and it makes it grey - not sure if that's the focus state or what but it seems weird to me. 
2. The count number based on selections is wrong. I should only count items and not folders! So in this example, it should say 2 out of 6 items, not 4 out of 9 items.

@mcarrano the badge is meant to show the number of items inside the folder.

**Comment by kmcfaul** [comment]
I think the grey background is from the hover state? And/or the focus state. It's definitely not 'selected'. @mcoker Any insight where the color is coming from? I think in TreeView the highlighting was to show placement with keyboard navigation but I'm not sure where the styling is coming from.

I'll work on updating the count number. I think I initially had it counting just the options (not folders) but it got a bit off in count and I changed it to all items. I'll see if I can figure it out.

@tlabaj @mcarrano I can update the behavior as such (toggle only expands, everything else checks the checkbox if it exists) we if want. Let me know!

**Comment by mcarrano** [comment]
I'll defer to @mmenestr on the checkbox and toggle interactions. I think users will figure it out either way.

**Comment by mmenestr** [comment]
Just looked at the tree view implantation and from what I can see, the only difference is that you can only expand by clicking the toggle which I'd be fine to copy for this component too.
When I click on the label in the tree view though, it doesn't get selected like you're saying @tlabaj so i'm confused! It only gets selected when I click on the checkbox.

**Comment by kmcfaul** [comment]
Updated. 
Only the toggle will expand/collapse a folder now.

**Comment by mcarrano** [comment]
@kmcfaul I think that the interaction makes sense now. My only issue it with the styling of selected items. Why do they have a gray background like this?

Screen Shot 2020-12-04 at 9 17 38 AM

I would only expect that the background shading is used to indicate hover or focus and selection in just indicated with the blue text. I don't have the core example to compare against. Can you check this @mcoker ?

**Comment by mmenestr** [comment]
@mcarrano that is the selected state for non-checkbox select elements, when you have a dual list with a single level, so it was just used again here. Alternatively, we could get rid of the added visual element, and only show a selection via the checkbox!

**Comment by mcarrano** [comment]
@mmenestr To me it feels redundant to have both.

**Comment by mmenestr** [comment]
@mcarrano ok, I don't have a strong preference either way, so I'm fine with removing all extra styling OR with just doing the blue text, with no grey background! I'll let you make the call :) The only reason for keeping it would be added emphasis - but I agree it may be redundant.

**Comment by mcarrano** [comment]
I think just having the checkbox selected and blue text is enough. The gray fill should be reserved to indicate hover/focus IMO.

**Review by mcarrano — APPROVED** [review]
Looks good pending core updates. Thanks @kmcfaul .

**Comment by kmcfaul** [comment]
Bumped core. There is a small issue that @mcoker is looking at (where the alignment of checkbox items with no children on the top level doesn't match other folders).

**Comment by jschuler** [comment]
@kmcfaul can you describe why you can't use the existing `<TreeView>` component? Seems a bit heavy to recreate it for dual list

**Comment by mcoker** [comment]
@jschuler I can say from the core side we talked to design and while they share a similar layout, they aren't the same component they may have different features, layouts, and don't necessarily need to change together when either is updated.

That said, I think there is enough design and functionality that maybe we have a shared component that both tree view and the dual list selector can use that give the base element and layout support, and those components modify the layout, state, etc changes unique to that component. wdyt?

**Comment by jschuler** [comment]
@mcoker can we not build in more features/variants/modifiers into the existing TreeView component? seems like it could cause more headaches in the future if they diverge already.

**Comment by kmcfaul** [comment]
Mostly what @mcoker said. The structure of the Tree component is similar but at least the Item is slightly different. The interaction patterns are also somewhat unique because DualListSelector is a specific use case of TreeView while TreeView is more generic. And the keyboard interaction (which needs to be defined for DualListSelector still) is entirely different since TreeView uses tabbing and DualListSelector ignores tabbing through the list entirely atm.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [inline packages/react-core/src/components/DualListSelector/DualListSelector.tsx:350]: seems this might be a bit brittle? considering you could have multiple branches with the same named nodes? clicking one would auto-check the other
Screen Shot 2020-12-08 at 10 06 39 AM

**Review by jschuler — CHANGES_REQUESTED** [review]

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:350]: Another example: In this case, selecting `Option Z` highlights both parent folders. Maybe this is an edge case but perhaps we shouldn't key off of text for identifying a node
Screen Shot 2020-12-08 at 10 12 36 AM

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:74]: this might be helpful
https://quicktype.io/typescript/

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [inline packages/react-core/src/components/DualListSelector/DualListSelector.tsx:540]: this is not i18n friendly (feel free to open followup issues as needed)

**Review by tlabaj — CHANGES_REQUESTED** [review]
Is the dual list selector tree only intended to be used with checks?  If not, it might be worth adding an example without checks and verifying that interactions work as expected.
cc @mcarrano

**Comment by mmenestr** [comment]
@tlabaj the dual list with **multiple levels** is only intended to be used with checkboxes yes.
But, a dual list with only a single level of options does not need checks though (ie: it's just a straight list of options with no nesting).

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:107]: I think I do need it because availableOptions can be re-passed in after a list update (like add/remove). To maintain the overall order I need an initial array.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:213]: It's making a copy to prevent array mutation.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:270]: They are similar in code but the arrays they target are different. I would personally rather keep them separate than have if/else logic inside to change around the targeted arrays, it seems more clearer to me this way. But am opening to change it if that's the consensus.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:326]: Ah yeah I see a couple instances where this could be shortened to `props.method && props.method(e)`. Will update.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:74]: availableOptions and chosenOptions both have `React.ReactNode[]` typing, I'm not sure what you mean here. Do you mean for me to change it to `React.ReactNode[] | DualListSelectorTreeItemData[]`?

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:350]: Good point. Maybe I should key off of `id` or a similar prop instead.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:540]: The custom string `availableOptionStatus` should overwrite either of the default strings. What about it isn't i18n friendly?

File: packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx:47]: Yeah I can rename it.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:107]: Would it be better to make a separate prop to use instead? Like `initialOptions`?

File: packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx:24]: I'm removing these props, since the Tree version isn't building a ref array (as of yet). They were holdovers from the initial DualListSelectorItem I worked off of.

Also, only `DualListSelector` is exported so I'm not sure if it matters, since a user can't access the sub-components?

**Comment by kmcfaul** [comment]
Updated with PR feedback.

- Tree now uses `id` for filtering/lookup etc (except in search function which still looks at text), to prevent duplicate selections when options share the same text.
- `checked` renamed to `isChecked`.
- Removed some unused props from the Tree/TreeItem. These were holdovers from the base DualListSelector code I started from, but are not used by the tree right now.
- Condensed some optional callback checking & triggering to inline.
- Re-enabled eslint no-unused props check.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:107]: oh, I am thinking that originalCopy is just a pointer to prop.availableOptions, so any changes to one should reflect in the other. Just would be cleaner not to have 3 things in code tracking availableOptions is all. But it's only a nit so if it causes issues it's okay the way it is
```
originalCopy
props.availableOptions
state.availableOptions
```

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:204]: array filter is already non-mutating and creates a new array. I'd go through and remove unnecessary usages of the map

Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:204]: btw, I often reference https://lorenstewart.me/2017/01/22/javascript-array-methods-mutating-vs-non-mutating/

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:74]: yes, so TS consumers can get hints and auto-completion as to what is supported in their options definition array when creating a tree configuration

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (jschuler) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:518]: oh sorry, i didn't see that

File: packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx
Comment (redallen) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx]: @tlabaj Not sure if this applies anymore, but the way to hide props is to start the comment with `@hide` like `/** @hide Internal field consumers need not worry about */`.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:74]: I'll also add this in the description for clarity.

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:204]: `filter` doesn't mutate, but I think it's because of the recursion that it's needed. I tried removing some of the `.map` but then I start losing options when it filters/checks/selects.

Comment (redallen) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:204]: @kmcfaul Can you use `Array.from(prevState.availableOptions)` instead? It's a little more explicit about the copy.

**Comment by tlabaj** [comment]
> @tlabaj the dual list with **multiple levels** is only intended to be used with checkboxes yes.
> But a dual list with only a single level of options does not need checks (ie: when it's just a straight list of options with no nesting).

@kmcfaul  if this is the case, is there really a need to have the `hasCheck` prop and all the logic associated with that?

**Review by jessiehuff — COMMENTED** [review]
For keyboard interaction, I think it would make the most sense to follow the [authoring practices for the treeview component](https://www.w3.org/TR/wai-aria-practices-1.2/#TreeView). I believe we should be following the pattern of a treeview that supports multi-selection, using left/right arrow keys to handle expansion, down/up arrow keys to handle focus between options, and Enter/Space for node selection. See the [keyboard interaction](https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-23) for all specifics. Also, I'm currently unable to use the keyboard to expand or select options.
Interestingly enough, I was able to perform some interactions with VO (checking the checkbox), but I was not able to expand or select options. I'm thinking that the interaction pattern between keyboard and screen reader might be linked here. I imagine this may be similar to [this open issue](https://github.com/patternfly/patternfly-react/issues/5149).

**Comment by kmcfaul** [comment]
@tlabaj If that's the case then I can remove that prop and always include checkboxes

**Comment by jschuler** [comment]
Can you also remove `hasChecks` from the DualListSelectorPane? And a rebase :)

**Comment by jessiehuff** [comment]
Added an [issue](https://github.com/patternfly/patternfly-react/issues/5231) for the keyboard/screen reader expectations. 🙂

File: packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx
Comment (tlabaj) [inline packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:21]: Shouldn't all the tree items have checks?  If so, this should not be needed.

Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:21]: I thought to leave it in just in case a user wanted to make something un-checkable (like a folder), but I can remove it.

Comment (tlabaj) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:21]: @mmenestr @mcarrano thoughts on this?  Do we want to allow them to not have the checkbox?

Comment (mcarrano) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:21]: I don't think so. It's not something I've seen. What do you think @mmenestr?

Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:21]: I'll remove it for now then.

**Review by tlabaj — APPROVED** [review]
Looks great!

**Comment by mmenestr** [comment]
Looks good! And like Matt said, every item should have a checkbox in this multi level dual list. The only thing i'd say now is that maybe we should revert the badge color back to the default grey for badges, since we no longer have the background hover state that would clash with it.

**Comment by nicolethoen** [comment]
There is a lot of conversation here, so this may be a done decision.
I'd just love some clarity.

In the tree view example, my mouse becomes a pointer and the item gets focus when I click on the text of an item, but nothing else happens. so it's focused, but not selected or checked? Why is that? Should clicking when the mouse looks like a pointer have a more obvious side effect?
Dec-09-2020 15-13-49


And i'm not sure if the enter key is doing anything? or the arrow keys? 
In the previous examples, you can tab to the list, then need to use the arrow keys to navigate the list items. So the tree list implementation is definitely different than the other examples. Is that on purpose?

Sorry if this is repeat.

**Comment by kmcfaul** [comment]
I'm not sure about the pointer. Perhaps it's a holdover when row selection did something and now I need to remove a class maybe? @mcoker Any thoughts?

Keyboard navigation has a follow up issue #5231, it needs to be defined and added.

**Comment by nicolethoen** [comment]
we are sure we don't want clicking the row to check the box?

**Comment by mcoker** [comment]
@nicolethoen @kmcfaul agree, the pointer cursor and background color change on hover should be paired with some kind of interactivity. It seems like we should be consistent between tree view and the dual list selector?

It seems like clicking anywhere in an item should perform the primary action for that item:

* Clicking a regular item selects it
* Clicking a regular expandable item expands it
* Clicking a checkbox item - should check it
* Clicking an expandable checkbox item - not sure? What's the primary action here?

wdyt @mcarrano?

We also made some structural updates to the tree view that I think we could make in this component, too, including making the text next to the checkbox a `<label>` (with a `for` attr that matches the `id` of the `input[type="check"]`), which will enable the checkbox to toggle the checked state when clicking on the label.

**Comment by mcarrano** [comment]
I agree with @mcoker above. Clicking anywhere on the item should select or de-select it by toggling the checkbox state. The only way to expand or collapse an item is to click on the expansion caret. I'm pretty sure that's consistent with what we said before. Can we make that change?

**Comment by kmcfaul** [comment]
Added the row click back in, toggling the checkbox (and therefore selection) just as the checkbox toggle would.

**Review by mcarrano — APPROVED** [review]
This is great. Thanks for making that change @kmcfaul !

**Review by mcoker — APPROVED** [review]
Looks super! 🤩

**Review by nicolethoen — APPROVED** [review]

**Comment by mmenestr** [comment]
before we merge @mcarrano, what do you think about changing the badge color to the lighter default grey? As opposed to the darker grey we initially switched so that it would not clash with the hover background (that we no longer have)

**Comment by mcarrano** [comment]
@mmenestr I think that we still do have the light gray background on hover (but not select). In any case, I'd like to get this merged this morning, if possible, so let's freeze the design for now. If there are follow on updates needed, they can be made in a future release since the component is still considered beta. Thanks.

**Review by tlabaj — APPROVED** [review]
LGTM

**Review by jschuler — APPROVED** [review]
LGTM

**Pull request merged**
